### DISCUSSION TITLE
Nodejs 18 Supporting

### DISCUSSION THREAD

**Discussion by strbaknxp** [FlowFuse/flowfuse#1876] [General]
Hi,

I would like to ask about the release timeline with bundling with nodejs.
Is some plan to release versions with supporting nodejs 18+ -> Node-Red 3.0.2 +?

Thank you!

**Comment by hardillb** [comment]
Hi,

Are we talking about the FlowForge hosted Cloud offering here or a self hosted instance?

For FlowForge Cloud we currently base our stacks on the `nodered/node-red:latest` docker containers which are based on NodeJS 16 and Node-RED 3.0.2. Once 3.1.0 is release then we will asses if it will make sense for the the 3.1.x based stacks to be based on NodeJS18 (the beta.2 stack is on the NodeJS16 currently)

If you are using a self hosted instance then you can always build your own stacks based on the Node-RED NodeJS 18 based containers.

**Reply by strbaknxp** [reply]
Oh, great! 
Thank you for quick response. :)

**Comment by kfernando2001** [comment]
Hi, Apologies if this is reptitive but when running Flow-fuse locally, it does not work on the latest version of node js (20) but does on 16? Is there any reason for this or am I doing something wrong?

**Reply by hardillb** [reply]
@kfernando2001 Can you supply more information about what "does not work"? What error are you seeing exactly?

**Reply by hardillb** [reply]
Just to clarify we build and run the core FlowFuse app on NodeJS 18 for the cloud version, but there shouldn't be any problem with running on NodeJS 20, so if you have a problem please raise an issue with details of how you are running FlowFuse.

**Reply by kfernando2001** [reply]
Hi, thanks for your response, and my apologies, it runs fine on NodeJS 20; the issue I was having was with 21. 
![Screenshot 2023-12-20 203935](https://github.com/FlowFuse/flowfuse/assets/153574113/d149ff67-1ca1-44b5-b4c1-484e89bfea08)
![Screenshot 2023-12-20 203908](https://github.com/FlowFuse/flowfuse/assets/153574113/519c9732-cba0-40a9-803c-390efe99f7aa)

**Reply by hardillb** [reply]
OK, that looks like a problem with a downstream dependency.

Just to check, you did do a fresh install after changing the NodeJS version? Because that looks to be a native module (for looking at CPU usage) so will have to be built for each different NodeJS/OS combination. npm will download or build the matching version at install time, so if you change the version after installing it can break things.

As a rule we will support LTS (even numbers) builds rather than the development builds (odd numbers)

**Reply by kfernando2001** [reply]
Hi sorry for the late reply, was just raising the issue, it works fine on NodeJS 20.

**Reply by kfernando2001** [reply]
Was hoping to ask another question; we are trying to figure out what advantages flow fuse adds to node-red. We know that node-red has various connection modules like MQTT/Modbus/Serial In, etc and flow fuse has device agent to manage node red on multiple devices, but what exactly does having the device agent on a device provide that node-red cant do itself? Any response would be greatly appreicated.

**Reply by hardillb** [reply]
FlowFuse adds a number of thing e.g:
- Remote access to the device editor if you don't have direct access to the network the device is running on
- access control
- Snapshots of flows (version control)
- The ability to push flows to multiple devices at once
- Remote logging

**Reply by kfernando2001** [reply]
Awesome, will test these out.
